compose/service.py

```diff
@@ -749,10 +749,12 @@ def custom_container_name(self):
         return self.options.get('container_name')
 
     def specifies_host_port(self):
-        for port in self.options.get('ports', []):
-            if ':' in str(port):
-                return True
-        return False
+        return any(
+            # Binding has an external port
+            not (isinstance(external_binding, tuple) and external_binding[1] is None)
+            for binding in self.options.get('ports', [])
+            for external_binding in (ports.split_port(binding)[1] or [])
+        )
```

Comment:
Using numeric indexes to split tuples is generally discouraged.

I think this could be improved slightly by splitting out some of the logic into a nested function

```python

def has_host_port(binding):
    _, external_binding = binding
    if not external_binding:
        return False
    ...


return any(has_host_port(binding) for binding in self.options.get('ports', []))
```